//source $JDE_DIR/Framework/common.sh;toBashDir $JDE_DIR JDE_BASH;
//cd $JDE_BASH/IotWebsocket/source/types/proto;webSource=$JDE_BASH/Public/src/web/proto;mklink FromServer.proto $webSource;mklink FromServer.pb.h $webSource;
//protoc --cpp_out . IotFromServer.proto

syntax = "proto3";
package Jde.Iot.FromServer;
import "google/protobuf/duration.proto";
import "google/protobuf/timestamp.proto";
import "FromServer.proto";
import "IotCommon.proto";

message MonitoredItemCreateResult{
  uint32 status_code=1;
  double revised_sampling_interval=2;
  uint32 revised_queue_size=3;
}

message SubscriptionAck{
	uint32 request_id=1;
	repeated MonitoredItemCreateResult results=2;
}

message Value{
	oneof of{
		bool boolean=1;
		uint32 byte=2;
		bytes byte_string=3;
		google.protobuf.Timestamp date=4;
		double double_value=5;
		google.protobuf.Duration duration=6;
		Jde.Iot.Proto.ExpandedNodeId expanded_node=7;
		float float_value=8;
		bytes guid=9;
		int32 int16=10;
		int32 int32=11;
		int64 int64=12;
		Jde.Iot.Proto.NodeId node=13;
		int32 sbyte=14;
		int32 status_code=15;
		string string_value=16;
		uint32 uint16=17;
		uint32 uint32=18;
		uint64 uint64=19;
		string xml_element=20;
	}
}

message NodeValues{
	string opc_id=1;
	Jde.Iot.Proto.ExpandedNodeId node=2;
	repeated Value values=3;
}

message UnsubscribeResult{
	uint32 request_id=1;
	repeated Jde.Iot.Proto.ExpandedNodeId successes=2;
	repeated Jde.Iot.Proto.ExpandedNodeId failures=3;
}

message MessageUnion{
	oneof Value{
		Jde.Web.FromServer.Acknowledgement acknowledgement=1;
		NodeValues node_values=2;
		Jde.Web.FromServer.Exception exception=3;
		SubscriptionAck subscription_ack=4;
		Jde.Web.FromServer.Complete complete=5;
		UnsubscribeResult unsubscribe_result=6;
		//QueryResult query=6;
		//Complete Complete=7;
	}
}

message Transmission{
	repeated MessageUnion messages=1;
}
